More SSL declarations

Aethylred opened the pull request:
added the following SSL declarations to vhost configuration

SSLOptions
SSLProtocol
SSLCipherSuite
SSLVerifyClient
SSLVerifyDepth

Set this up so that if a default value (as per http://httpd.apache.org/docs/2.2/mod/mod_ssl.html ) is used, it will be omitted.

This should be sufficient for setting up puppetmaster on `apache::mod::passenger`

Commit:
Add parameter for purging vdir

The new parameter $purge_vdir in apache class gives the possibility to disable the purging of the virtual hosts directory.

blkperl committed:
Enable puppet 3.0.1 in travis.yml

A fix for rodjek/rspec-puppet#58 has been merged which
should allow puppet 3.0.1 to run the spec tests

kbarber committed:
Build all branches on travis

We should build all branches for travis, otherwise users won't be able to test
their own feature branches.

Signed-off-by: Ken Barber <ken@bob.sh>

kbarber committed:
Merge branch 'update_travis'

* update_travis:
  Build all branches on travis
  Enable puppet 3.0.1 in travis.yml

blkperl committed:
Standardize travis and gemfile

This commit standardizes the travis and gemfile to be similar
to the files in stdlib.

kbarber committed:
Merge pull request #144 from blkperl/update_travis

Standardize travis and gemfile

richardc committed:
Restrict the versions and add 3.1

~> 2.6 will match 2.7

kbarber committed:
Merge pull request #159 from richardc/patch-1

Restrict the versions and add 3.1

Commit:
added the following SSL declarations to vhost configuration

SSLOptions
SSLProtocol
SSLCipherSuite
SSLVerifyClient
SSLVerifyDepth

Set this up so that if a default value (as per http://httpd.apache.org/docs/2.2/mod/mod_ssl.html ) is used, it will be omitted.

Commit:
the SSLCACertificatePath declaration has to be overridable, and is not always in the same place as ssl_certs_dir

hunner commented on templates/vhost/_ssl.erb:
```diff
@@ -7,7 +7,7 @@
   SSLCertificateChainFile <%= @ssl_chain %>
 <% end -%>
-  SSLCACertificatePath    <%= @ssl_certs_dir %>
+  SSLCACertificatePath    <%= @ssl_ca_dir %>
```
Currently the `ssl_certs_dir` parameter for `apache::vhost` is used to set the `SSLCACertificatePath` and nothing else. What else do you envision the `ssl_certs_dir` parameter to be specifying?

Aethylred replied:
I've seen setups where all the certs and keys are lumped together in one directory, but more commonly I've seen things with separate subdirectories for public certificates, private keys, _and_ the CA certificates. I don't think we could plan the module to handle all situations, it'd be better to set it up one way and handle the exceptions as they come in as issues.

In this case, the CA certificates are not stored in the same place as `$ssl_cert_path` (i.e. not with the certificates or the keys...) and were installed by the puppetmaster-passenger package so relocating them would be complicated...

What I'd really like to see in Puppet is an x509 certificate module that behaves like [`ssh::auth`](http://projects.puppetlabs.com/projects/1/wiki/Module_Ssh_Auth_Patterns) so that the puppetmaster generates and stores self signed certificates, and then deploys them to nodes, but that's outside the scope of this module.

hunner commented on manifests/vhost.pp:
```diff
@@ -64,9 +64,15 @@
     $ssl_crl            = $apache::default_ssl_crl,
     $ssl_certs_dir      = $apache::params::ssl_certs_dir,
+    $ssl_protocol       = 'all',
+    $ssl_cipher_suite   = 'ALL:!ADH:RC4+RSA:+HIGH:+MEDIUM:+LOW:+SSLv2:+EXP',
```
The cipher suite is currently statically set in the template for `apache::mod::ssl` at https://github.com/hunner/puppetlabs-apache/blob/refactor_module/templates/mod/ssl.conf.erb#L19 . It is configured with a cipher suite that is recommended by the Apache foundation as seen in http://svn.apache.org/viewvc?view=revision&revision=1203752

If you would like to override this on a per-vhost basis, I would prefer that the default be `HIGH:MEDIUM:!aNULL:!MD5` set in `apache::params` and used by both the `apache::mod::ssl` and `apache::vhost` manifests.

Aethylred replied:
The default I've chosen is simply the default value for `mod_ssl`, so that when it matches in the template, no line is written. If the user supplies a less secure `SSLOptions` string then it's their configuration mistake, not an issue with out module deploying that value.

hunner commented on templates/vhost/_ssl.erb:
```diff
@@ -17,7 +17,19 @@
+<% end -%>
   <FilesMatch "\.(cgi|shtml|phtml|php)$">
-    SSLOptions +StdEnvVars
+    SSLOptions <%= @ssl_options %>
```
In this directive, `StdEnvVars` is only enabled on CGI/SSI requests as per http://httpd.apache.org/docs/2.2/mod/mod_ssl.html#ssloptions .

Perhaps we should leave this `SSLOptions` directive as-is, and instead allow it to be set in the `VirtualHost` declaration instead of inside the `FilesMatch` declaration?

Aethylred replied:
Yes, that was how I had implemented it before, but chose to conform to your implementation.

hunner commented:
Hey @Aethylred, thanks for this pull request! It has a bunch of useful parameters in it.

One comment about setting defaults is that we prefer to use `undef` or empty string/array defaults so that if upstream defaults change (such as for Apache's httpd), they can still be accounted for without having to change the code in the module. Then you can just say things like `<% if @ssl_cipher_suite %>` in your template without checking for a specific value.

I'll get some rspec tests written for this and get it ready for merging. Thanks!

Aethylred commented:
will consider using `undef` more often. I think the setting the defaults values like I have is safe with conservative applications like Apache, but perhaps simpler behavior might be preferred.

Commit:
SSLOptions needs to be for the general VirtualHost block, not wrapped in a filematch.

...puppetmaster under passenger fails without this.

Commit:
@ is the correct variable decorator for ssl_options

bodepd committed:
Merge pull request #139 from mjanser/patch-1

Add parameter for purging vdir

Commit:
Capitalize RedHat to be in-line with actual factor output

Commit:
Fix vhost defined type unit test context to reflect actual test being performed.

context lists 'Debian' but test is for RedHat.

adrienthebo committed:
Merge pull request #171 from wolfspyre/master

Small fix to unit tests

Commit:
release 0.6.0

bodepd committed:
Merge pull request #174 from bodepd/0_6_0

release 0.6.0

trlinkin committed:
Custom logging for apache::vhost

This patch adds the ability to change how vhost logging works. This patch
adds new parameters to 'apache::vhost' define type. This allows users to
specify a log format or nickname for the 'CustomLog' directive. This also
allows users to specify a pipe as the destination for the 'ErrorLog' and
'CustomLog' directives.

hunner commented:
I am going to get to this and #6, I just am trying to get the baseline of the module established before adding additional parts where possible! Thanks for being patient.

hunner committed:
Merge branch 'refactor_module' into master

Conflicts:
	.travis.yml
	manifests/init.pp
	spec/defines/mod_spec.rb
	spec/defines/vhost_spec.rb

Aethylred commented:
Thanks Hunner, I've been using my fork of your fork with my additions merged in. I've also picked up some rspec, but won't be able to get back to these for a while.

trlinkin committed:
Spec tests for apache::vhost logging improvements

The enhancements to vhost logging are currently passing the previous
spec tests in place. However the enhancements added two new ways the
apache::vhost type should fail. The enhancements also added some extra
logic surrounding the CustomLog and ErrorLog directives.

This patch extends the suite of tests already running against
apache::vhost logging enhancements added in 4226690eb3. The new
tests verify the correct functionality of the logging enhancements.

trlinkin committed:
Correct regexp warning in vhost_spec.rb

Without this patch, running the spec tests encounter a warning
about a regular expresion with a '}' that has not been escaped. This
warning is from line 238 in the vhost_spec.rb file.

This commit eliminates that warning.

trlinkin committed:
Add documentation for apache::vhost logging

Commit 6f4a317 added new paramters for controlling logging for the
apache::vhost type. This commit adds documentation for those changes.

hunner committed:
Merge pull request #183 from trlinkin/master

Remove spec test warning

trlinkin committed:
Refactor apache::vhost log file destination logic

This commit combines the blocks of logic determining the apache::vhost
logging values.

hunner committed:
Merge pull request #184 from trlinkin/refactor_vhost_logformat

Refactor vhost logformat

trlinkin committed:
Add Validation and rspec tests for $error_log

Without this commit, there is no validation or testing for the
$error_log parameter in the apache::vhost type.

This commit add boolean validation to the apache::vhost type. It also
adds two rspec tests to test the proper function of the parameter.

Commit:
Merge branch 'master' into more_ssl

Conflicts:
	manifests/vhost.pp

Commit:
Change from default values to using undef

hunner committed:
Merge pull request #186 from trlinkin/refactor_vhost_logformat

Add Validation and rspec tests for $error_log

Aethylred commented:
ok, a bit messy, as I've merged my branch with the new master.

nanliu committed:
Fix mod::prefork dependencies

A dependency on the package resolves the following error:
"Error: /Stage[main]/Apache::Mod::Prefork/File_line[/etc/sysconfig/httpd
prefork enable]: Could not evaluate: No such file or directory -
/etc/sysconfig/httpd"

trlinkin committed:
Move mpm_worker 'ServerLimit' line in ERB

It turns out that the order of MPM directives is important in certain
cases. Without this commit, the 'ServerLimit' directive was being
processed too late to be used in calculations for 'ThreadsPerChild'.

This commit moves the 'ServerLimit' directive to the top of the
directives in the ERB. This was based on Apache documentation and an
example config of theirs.

hunner committed:
Merge pull request #191 from trlinkin/master

Move 'ServerLimit' line in worker.erb

tjikkun committed:
Add proxy_pass parameter

mxey committed:
Use correct template for dav_fs module

hunner committed:
Merge pull request #188 from nanliu/prefork

Fix mod::prefork dependencies

hunner committed:
Merge pull request #194 from mxey/master

Use correct template for dav_fs module

hunner committed:
Merge pull request #193 from oxilion/proxy_pass

Add proxy_pass parameter

pronix committed:
fix variable for latest puppet version

apenney committed:
Minor tweak to set this to false as it's a replaced parameter.

hunner committed:
Merge pull request #209 from apenney/vdirpurgefix

Minor tweak to set this to false as it's a replaced parameter.

hunner committed:
Merge pull request #208 from pronix/master

fix variables for latest puppet

Commit:
Merge branch 'master' into more_ssl
